### ISSUE TITLE
Fix expand/collapse

### ISSUE THREAD

**Issue by hidde** [w3c/wai-wcag-supporting-documents-redesign#55]
When we use `h2` (or other headings) inside `summary` elements, some screenreaders, including JAWS, will no longer recognise the heading as a heading, making “navigate by heading”.

Question: is this somehting we should fix or is it a bug only in old version of Jaws and we could decide to leave?

To solve that, we'll write JavaScript that: 

* recognises headings inside `summary`/`details`
* duplicates the heading just before the `details` element, with `visually-hidden` class
* sets the actual heading to `role="presentation"` 

So from: 

```html
<details>
  <summary><h2>Heading</h2></summary>
  Content
</details>
```

we change to:

```html
<h2 class="visually-hidden">Heading</h2>
<details>
  <summary><h2 role="presentation">Heading</h2></summary>
  Content
</details>
```

## Caveats

We'll remove links from _headings_ in expand/collapses, because we'll assume semantics are removed as we are setting `role="presentation"`.

We should assume no semantic content will work if it is inside the heading that ends up getting `role=presentation`.

## Other tasks

Also: add documentation in [Ex/col](https://wai-website-theme.netlify.app/components/excol/single/): explain that we did this fix, also explain 'don't add semantic elements inside heading'.

**Comment by shawna-slh** [comment]
This is a site-side issue. Steve will look at and maybe check with @daniel-montalvo

**Comment by SteveALee** [comment]
Here's [an example using the collapsable/heading](https://w3c.github.io/wai-wcag-supporting-documents-redesign/2020-10-13-all.html)

I prefer a javascript free solution so a solution might be:

- normal heading
- an expand button to show/hide the list 

A modified version of this can be seen in the [latest supplemental guidance](https://618aa19c8a704c0007cadb84--wai-wcag-supplemental.netlify.app/wcag-supplemental/supplemental-guidance/) where I added always visible summary info and a link per section :

- normal heading of Objective name
- summary text -
- a link to the page with dull Objective text 
- an expand button to show the list of child patterns

Questions on this include:

- how is visual clarity giving the heading styling?
- is screen reader experience good given?
  - link text is long and repeats heading so they are unique on page
  - the button text is non unique

**Comment by SteveALee** [comment]
@daniel-montalvo would you be able to review the above comment from a screen reader users perspective? There's background info in the issue text at the top.

Thanks

**Comment by daniel-montalvo** [comment]
Thanks @hidde  @SteveALee  and @shawna-slh  for your work on this.

I very much like this approach.

> * is screen reader experience good given?
>   * link text is long and repeats heading so they are unique on page
>   * the button text is non unique

Couldn't we add the associated pattern after the "Show pattern" string? In this example there are eight "Show pattern" buttons. That makes it a bit hard to parse. Something like "Show patterns for Help Users Find What They Need" would help.

**Comment by shawna-slh** [comment]
Thanks for working on this, Steve.

I think we want to be consistent throughout the WAI website, unless there is a strong reason to do things differently -- in which case, then we should! :-)

Expand-collapse is used in many places currrently, e.g.:
* https://www.w3.org/WAI/fundamentals/accessibility-principles/
* https://www.w3.org/WAI/people-use-web/user-stories/
* https://www.w3.org/WAI/people-use-web/abilities-barriers/
* https://www.w3.org/WAI/people-use-web/tools-techniques/
* https://www.w3.org/WAI/planning-and-managing/initiate/
* https://www.w3.org/WAI/planning-and-managing/plan/
* https://www.w3.org/WAI/planning-and-managing/implement/
* https://www.w3.org/WAI/planning-and-managing/sustain/
* https://www.w3.org/WAI/planning/statements/
* https://www.w3.org/WAI/test-evaluate/preliminary/
* https://www.w3.org/WAI/teach-advocate/accessible-presentations/
* https://www.w3.org/WAI/resources/
* https://www.w3.org/WAI/curricula/developer-modules/page-structure/ and most of the other curricula pages

I think replacing the heading with additional wording "Show xyz" adds unnecessary (even annoying) visual complexity.

Let's chat about ideas...

Let's chat

**Comment by SteveALee** [comment]
Agree should be sitewide consistency. There are more case in the WCAG supporting docs redesign project.

With respect to lack of visual affordance that this heading is actually interactive and a disclose I see examples that add little icons, for example [this page](https://www.dementiauk.org/get-involved/campaigns/guiding-the-way/)

**Comment by SteveALee** [comment]
See this [preview of supplemental](https://619622708365e0000810ef64--wai-wcag-supplemental.netlify.app/wcag-supplemental/supplemental-guidance/) for a suggestion for the new pages. It changes the headings to list items and hides the marker.

**Comment by daniel-montalvo** [comment]
* List and list items are still not picked up by JAWS. It seems they ignore all the semantics inside a `summary` element.
* Also the "link inside a button" seems to be a little bit of a convoluted pattern. For example, "Help Users Understand What Things are and How to Use Them" is picked up as a button and, when expanded, is picked up as a link by NVDA. This is confusing.
  * With NVDA, you have to tab to the link to be able to press it (it is not enough with down arrowing to it)
  * With JAWS you cannot press the "Help Users Understand What Things are and How to Use Them" link. You can only access the links in the second level list once you expand the first level, due to JAWS ignoring semantics inside the `summary` element.

**Comment by SteveALee** [comment]
@daniel-montalvo 

> List and list items are still not picked up by JAWS.

Sigh! Why are they so broken with basics. 

> Also the "link inside a button" seems to be a little bit of a convoluted pattern. For example, "Help Users Understand What Things are and How to Use Them" is picked up as a button and, when expanded, is picked up as a link

I don't get that, it should only be links. I'll could see if the JS for the "custom collapse" is messing with the DOM. 

BUT after discussion with @Shawn we're going to change over to just lists and  no "custom collapse" at all as it is different and newer evergreen browsers show little visual "markers" for collapsed state.. That should fix the problems BUT we need to implement a Collapse/Expand All feature.

**Comment by SteveALee** [comment]
@daniel-montalvo So the [HTML spec for details](https://html.spec.whatwg.org/multipage/interactive-elements.html#the-details-element) says content can be controls but only says "show" and "hide" I don't know the spec well enough to know if hide means "display:none" so not in ally tree or they just dropped that in the spec? Do you have any insight?

**Comment by daniel-montalvo** [comment]
My understanding is that the HTML spec allows what we are doing. Although it may not be sufficiently understood by users and may not meet their expectations.

Rationale behind JAWS ignoring semantics  inside `summary` element might be in [HTML Accessibility API Mappings 1.0](https://www.w3.org/TR/html-aam-1.0/#summary-element)

**Comment by SteveALee** [comment]
So now I rad the thing, the spec is clear, the `summary` element's content model is '[phrasing content](https://html.spec.whatwg.org/multipage/dom.html#phrasing-content-2)' and that does NOT include links! 

So Jaws is OK and we need to rethink having a link in the summary :(

**Comment by SteveALee** [comment]
@daniel-montalvo @shawna-slh 

I've created a new [details test page](https://deploy-preview-36--wai-wcag-supplemental.netlify.app/wcag-supplemental/details-test/) which I hope might work visually and with a SR. I stripped out all the theme JS and CSS that messes about with the `details` and `summary` elements so we know what we are detailing with.

The markup is an `a` in the `summary` followed by and `ul`, like so

```
<details>
  <summary>
      <a href="#top1">Top level 1</a>
  </summary>
  <ul>
    <li><a href="#child11">Child 1.1</a></li>
    <li><a href="#child12">Child 1.2</a></li>
  </ul>
</details>
<details>
  <summary>
      <a href="#top2">Top level 2</a>
  </summary>
  <ul>
    <li><a href="#child21">Child 2.1</a></li>
    <li><a href="#child22">Child 2.2</a></li>
  </ul>
</details>
```

**Comment by shawna-slh** [comment]
There are two issues:
1. What we need to do for the expand-collapse used throughout the WAI website.
2. What to do for the All Supplemental Guidance page. I think that can be a standard tree view?

**Comment by SteveALee** [comment]
> 1.    What we need to do for the expand-collapse used throughout the WAI website.

Isn't this working OK as is with the `include excol`? Except if we want Headers in the summary. The spec says Headers are "Optional" so seem indeterminate. D'oh.

> 2.   What to do for the All Supplemental Guidance page. I think that can be a standard tree view?

Does the above work? I think this issue is for supporting docs/supplemental only. 
I'm not sure what a "a standard tree view" as there's no such HTML element. Does the  above code meet you idea or nested lists only with no collapse?

**Comment by shawna-slh** [comment]
> Does the above work?

Yes, nested list with expand collapse twisty/arrow good!

**Comment by daniel-montalvo** [comment]
Hey @SteveALee @shawna-slh 

[This details page](https://deploy-preview-36--wai-wcag-supplemental.netlify.app/wcag-supplemental/details-test/) works better. Note that JAWS detects the button role but omits the link role. If there is additional keyboard interaction expected inside the `summary` element, for example expanding the content and then clicking a link, chances are JAWS users will still miss that. Happy to get on the phone for this if needed.

**Comment by SteveALee** [comment]
@daniel-montalvo sigh!

They must be mapping the content and breaking it in the process.

1 - there is no real button - just the select interactive "marker" - I guess they need to present it as something 
2 - a link in summary IS valid content and they are losing it 

This makes the per Objective content difficult to navigate to. The hack is to a pattern and then to the objective from there.

So can we find if there is a bug at JAWS and when they plan to fix it? Do they even have a public issue tracker?
Should we work around this or not?

**Comment by shawna-slh** [comment]
I think the implementation that we're using for WCAG materials is OK.

There is still a site-wide issue https://github.com/w3c/wai-website/issues/280

Steve, Daniel, others, feel free to comment more here or there. Thanks.
